dcos-log/journal/reader/read.go

```diff
@@ -219,3 +220,68 @@ func (r *Reader) Close() error {
 	}
 	return r.Journal.Close()
 }
+
+// Follow is a wrapper function, which can be called multiple times to mimic a journal tailing.
+func (r *Reader) Follow(wait time.Duration, writer io.Writer) error {
+	n, err := io.Copy(writer, r)
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	// if the number of read lines more then 0, we did not reach the journald bottom and can exit early
+	if n > 0 {
+		return nil
+	}
+
+	// if we reached the journald bottom, we'll have to wait and learn the current status of journald
+	// SD_JOURNAL_INVALIDATE indicates that the journald files were removed from the filesystem and now we need to close
+	// the opened files handlers and reopened with original user parameters.
+	// https://www.freedesktop.org/software/systemd/man/sd_journal_get_fd.html#Return%20Value
+	if r.Journal.Wait(wait) == sdjournal.SD_JOURNAL_INVALIDATE {
+		logrus.Infof("SD_JOURNAL_INVALIDATE, reopened journal")
```

Comment:
It might be helpful to log which journald file this is.

Reply:
unfortunately this is not possible with systemd